src/event-cache/__tests__/EventCache.test.ts

```diff
@@ -158,6 +357,98 @@ describe('EventCache tests', () => {
+                metadata: `{"version":"1.0.0","aws:client":"${INSTALL_MODULE}","aws:clientVersion":"${WEB_CLIENT_VERSION}"}`,
+                details: '{}'
+            },
+            {
+                id: expect.stringMatching(/[0-9a-f\-]+/),
+                timestamp: new Date(),
+                type: EVENT3_SCHEMA,
+                metadata: `{"version":"1.0.0","aws:client":"${INSTALL_MODULE}","aws:clientVersion":"${WEB_CLIENT_VERSION}"}`,
+                details: '{}'
+            },
+            {
+                id: expect.stringMatching(/[0-9a-f\-]+/),
+                timestamp: new Date(),
+                type: EVENT2_SCHEMA,
+                metadata: `{"version":"1.0.0","aws:client":"${INSTALL_MODULE}","aws:clientVersion":"${WEB_CLIENT_VERSION}"}`,
+                details: '{}'
+            }
+        ];
+
+        // Run
+        // Candidates 1 and 3 should show up before 2
+        eventCache.recordEvent(EVENT2_SCHEMA, {}, { isCandidate: false });
+        eventCache.recordEvent(EVENT1_SCHEMA, {}, { isCandidate: true });
+        eventCache.recordEvent(EVENT3_SCHEMA, {}, { isCandidate: true });
+        const eventBatch: RumEvent[] = eventCache.getEventBatch(true);
+
+        // Assert
+        expect(eventBatch).toEqual(expect.arrayContaining(expectedEvents));
+    });
+
+    test('getEventBatch does not drop candidates or regular events that remain in the cache', async () => {
+        // Init
+        const EVENT1_SCHEMA = 'com.amazon.rum.event1';
+        const EVENT2_SCHEMA = 'com.amazon.rum.event2';
+        const eventCache: EventCache = Utils.createEventCache({
+            ...DEFAULT_CONFIG,
+            ...{
+                batchLimit: 3,
+                eventCacheSize: 5,
+                candidatesCacheSize: 5
+            }
+        });
+
+        // Run
+        // Candidates 1 and 3 should show up before 2
+        for (let i = 0; i < 5; i++) {
+            eventCache.recordEvent(EVENT1_SCHEMA, {}, { isCandidate: true });
+            eventCache.recordEvent(EVENT2_SCHEMA, {}, { isCandidate: false });
+        }
+
+        // Assert
+        // Flush candidates
+        expect(eventCache.getEventBatch(true).map((x) => x.type)).toEqual(
+            expect.arrayContaining([
+                EVENT1_SCHEMA,
+                EVENT1_SCHEMA,
+                EVENT1_SCHEMA
+            ])
+        );
+        // Do a regular batch
+        expect(eventCache.getEventBatch(false).map((x) => x.type)).toEqual(
+            expect.arrayContaining([
+                EVENT2_SCHEMA,
+                EVENT2_SCHEMA,
+                EVENT2_SCHEMA
+            ])
+        );
+        // Flush candidates again
+        expect(eventCache.getEventBatch(true).map((x) => x.type)).toEqual(
+            expect.arrayContaining([
+                EVENT1_SCHEMA,
+                EVENT1_SCHEMA,
+                EVENT2_SCHEMA
+            ])
+        );
+        // Flush candidates, but there are no remaining
```

Comment:
question: do you mean no more events to be flushed? the cache still has one instance of event_2?

Reply:
refactored in rev2